fix: Use ascending sort for percentile

matthewfeickert opened the pull request:
Resolves #864

This changes the behavior of `percentile` and `quantiles`, as previously the sort order had been descending. This is being changed as it differs in behavior from [NumPy's `percentile` function](https://numpy.org/doc/1.18/reference/generated/numpy.percentile.html) and the docstring describes behavior that is consistent with NumPy's implementation.

The main change occurs replacing `_sort_tensor` with a single call to [`tf.sort`](https://www.tensorflow.org/versions/r2.1/api_docs/python/tf/sort), which is then used in `percentile`. In `_sort_tensor` [`tf.math.top_k`](https://www.tensorflow.org/versions/r2.1/api_docs/python/tf/math/top_k) (which results in descending sort) was used. When swapped for [`tf.sort`](https://www.tensorflow.org/versions/r2.1/api_docs/python/tf/sort) (with the default values of `k=-1, direction='ASCENDING'` explicitly passed to make the behavior more clear to a reader) there is no longer a need to ensure that the output is reshaped properly. As `_sort_tensor` is now just a single function call it is then removed in favor of `tf.sort`. This then necessitates the change in definition of `frac_at_q_or_above` such that it is renamed to `frac_at_q_or_above` and the swap of `tf.math.ceil` and `tf.math.floor`.

This could have also been achieved with

```python
sorted_, _ = tf.math.top_k(tensor, k=tf.shape(tensor)[-1])
sorted_ = tf.reverse(sorted_, [-1])
```

but as there doesn't seem to be a clear reason to use `tf.math.top_k` over `tf.sort`, one function call seemed preferable to two.

A test is also added that validates the edge case discussed in Issue #864 that prompted this PR.

cc @lukasheinrich @kratsg

matthewfeickert commented:
@jvdillon @csuter As you discussed this issue with me in Issue #864 I thought I would tag you here in the event that it might be easier for you two to review this PR. I have no idea how the TFP dev team does PR review so my apologies if there is a designated review shifter that will get around to this at some point.

I also don't see any CI getting kicked off(?) so if it would help to provide the test script that I've been running locally as well to check all the docstring examples for `percentile` and `quantiles` just let me know and I'm happy to provide that as well in the PR discussion.

ColCarroll commented on tensorflow_probability/python/stats/quantiles.py:
```diff
@@ -548,23 +548,20 @@ def percentile(x,
 
-    # Sort everything, not just the top 'k' entries, which allows multiple calls
-    # to sort only once (under the hood) and use CSE.
+    # Sort (in ascending order) everything, not just the top 'k' entries, which
```
you can delete `not just the top k entries`, since that was referring to the `_sort_tensor` implementation.

matthewfeickert replied:
Done.

matthewfeickert commented on tensorflow_probability/python/stats/quantiles.py:
```diff
@@ -548,23 +548,20 @@ def percentile(x,
       y = _move_dims_to_flat_end(x, axis, x_ndims, right_end=True)
 
-    frac_at_q_or_above = 1. - q / 100.
+    frac_at_q_or_above = q / 100.
```
Good catch!

matthewfeickert commented:
@ColCarroll Thanks very much for the review! I've gone ahead and implemented those changes (also thanks for catching them!) and rebased them into the first commit when I rebased against upstream.

(My apologies if this is not your preferred method of changes during code review. I just figured since TFP doesn't do squash and merge PRs this would be cleaner to track the relevant changes for whenever it gets moved into a release than if it was spread across a few extra cleaning commits.)

csuter commented on tensorflow_probability/python/stats/quantiles.py:
```diff
@@ -899,7 +896,8 @@ def _move_dims_to_flat_end(x, axis, x_ndims, right_end=True):
+  """Use `sort` to sort a `Tensor` in ascending order along
+     the last dimension."""
+  sorted_ = tf.sort(tensor, axis=-1, direction='ASCENDING')
   tensorshape_util.set_shape(sorted_, tensor.shape)
```
i think the set_shape is no longer needed, since tf.sort will not change the shape (top_k would have the potential to, although we were abusing it slightly by setting k to the size of the input).

matthewfeickert replied:
@csuter As this can then be done in 1 line now
```python
return tf.sort(tensor, axis=-1, direction='ASCENDING')
```
and there is only 1 function call to `_sort_tensor` should I just remove `_sort_tensor` altogether and instead have
```python
sorted_y = tf.sort(y, axis=-1, direction='ASCENDING')
```
?

csuter replied:
yes, that sounds great. thanks!

matthewfeickert replied:
Done.

matthewfeickert commented:
@srvasude Now that this has been reviewed and approved, can you explain what are the conditions that are required before this can get merged? I hope that I'm not coming across as impatient or demanding, I'm just unfamiliar with the how contributions to TFP slot in with development from the core dev team.

srvasude commented:
Hi, apologies for the slowness. The process is for us to pull in the request, and let it go through internal review before pushing. I was in the process of this, but was coming across some issues in our internal testing. In particular the percentile tests (specifically the test class PercentileTestWithLinearInterpolation in quantiles_test) is failing. Do you mind checking if any changes need to be made to quantile / quantile_test now that the sort order has been changed?

matthewfeickert commented:
> Hi, apologies for the slowness. The process is for us to pull in the request, and let it go through internal review before pushing. 

Nothing to apologize about, and this information is actually very helpful for understanding how things work in general. Thank you for sharing it.

> I was in the process of this, but was coming across some issues in our internal testing. In particular the percentile tests (specifically the test class PercentileTestWithLinearInterpolation in quantiles_test) is failing. Do you mind checking if any changes need to be made to quantile / quantile_test now that the sort order has been changed?

Sure. Luckily the changes that I've made are incredibly minor and only show up in the test case that I've added in the test that I added to `quantile_test`. I have a test script that I've been running locally as well that runs that test and additionally checks all the docstring examples for `percentile` and `quantiles`. I've run it each time I rebase this PR of `master` and things are passing. If this would be helpful I'm more than happy to provide it.

matthewfeickert commented:
@srvasude Now that the CI is back up and running I see that there is indeed a deviation between NumPy and my branch for `axis=None`. I've moved this PR into draft mode until I fix this. I'll ping you for review once I get this fixed and fully passing CI. I apologize for not catching this before I requested a review in the first place.

matthewfeickert commented:
@srvasude This is passing CI and is ready for review now. 

What was causing the errors in the tests was that I had missed that given my change in how 'lower' and 'upper' were defined, I forgot to swap their usage as well &mdash; sloppy mistake on my part. I'm sorry about that error, but I've rebased this in against `master` and with that fixed and this passing CI I think it is ready for final review.

matthewfeickert commented:
Excellent! Thank you very much @srvasude for the review and helping me get this PR done &mdash; I really appreciate it! Also thank you @csuter and @ColCarroll for your reviews and input which made all of this easier. :)

csuter commented:
Thanks for your contribution, Matthew. It's been a pleasure to work with you! :)
